python/uv/private/lock.bzl

```diff
@@ -13,113 +13,424 @@
+      to the same command that would be run in the `name` action. This will
+      update the source copy of the requirements file.
+    - `name.update`: a target that can be run to update the source-tree version
+      of the requirements lock file. The output can be fed to the
+      {obj}`pip.parse` bzlmod extension tag class. Note, you can use
+      `native_test` to wrap this target to make a test.
+
+    Args:
+        name: {type}`str` The prefix of all targets created by this macro.
+        srcs: {type}`list[Label]` The sources that will be used. Add all of the
+            files that would be passed as srcs to the `uv pip compile` command.
+        out: {type}`str` The output file relative to the package.
+        args: {type}`list[str]` The list of args to pass to uv. Note, these are
+            written into the runnable `name.run` target.
+        env: {type}`dict[str, str]` the environment variables to set. Note, this
+            is passed as is and the environment variables are not expanded.
+        build_constraints: {type}`list[Label]` The list of build constraints to use.
+        constraints: {type}`list[Label]` The list of constraints files to use.
+        generate_hashes: {type}`bool` Generate hashes for all of the
+            requirements. This is a must if you want to use
+            {attr}`pip.parse.experimental_index_url`. Defaults to `True`.
+        strip_extras: {type}`bool` whether to strip extras from the output.
+            Currently `rules_python` requires `--no-strip-extras` to properly
+            function, but sometimes one may want to not have the extras if you
+            are compiling the requirements file for using it as a constraints
+            file. Defaults to `False`.
+        python_version: {type}`str | None` the python_version to transition to
+            when locking the requirements.
+        **kwargs: Extra kwargs passed to the `name.update` {obj}`py_binary` target.
+    """
+    update_target = "{}.update".format(name)
+    locker_target = "{}.run".format(name)
+    template_target = "_{}.update_template".format(name)
 
     # Check if the output file already exists, if yes, first copy it to the
     # output file location in order to make `uv` not change the requirements if
     # we are just running the command.
-    if native.glob([out]):
-        cmd = "cp -v $(location {}) $@; {}".format(out, cmd)
-        srcs.append(out)
+    maybe_out = _maybe_file(out)
+    target_compatible_with = kwargs.pop("target_compatible_with", None)
+    if target_compatible_with == None and not BZLMOD_ENABLED:
+        target_compatible_with = ["@platforms//:incompatible"]
+    elif target_compatible_with == None:
+        target_compatible_with = []
 
-    native.genrule(
+    # FIXME @aignas 2025-03-17: should we have one more target that transitions
+    # the python_version to ensure that if somebody calls `bazel build
+    # :requirements` that it is locked with the right `python_version`?
```

Comment:
A separate target, no. An attribute with rule-level cfg transition, yes.

This also enables tricks like this:

```
lock(python_version="3.10", srcs=select(":py310": "requirements_310.txt", ...))
```

Similarly, because an `attr.output` is not used, the outputs can be varied to e.g. include the python version (or platform, etc) into the output file name.